README.md
[![Build Status](https://travis-ci.org/evertrue/et_cassandra-cookbook.svg?branch=master)](https://travis-ci.org/evertrue/et_cassandra-cookbook)

# et_cassandra

This cookbook will install and configure a Cassandra node, optionally allowing for the installation of OpsCenter and the Datastax Agent.


CHANGELOG.md
# CHANGELOG for et_cassandra

## v7.1.15 (2021-09-30)

* switching to apt install s3cmd

## [7.1.14] (2021-03-11)

* Update pip url

## [7.1.13] (2021-02-28)

* Specify pip url and pin pip & virtualenv versions

## [7.1.10] (2020-02-11)

* switched to poise-python

## [v7.1.9] (2019-03-07)

* Set node java_home value

## [v7.1.8] (2017-10-30)

* Set JAVA_HOME for root

## [v7.1.7] (2017-10-09)

* Switch to OpenJDK from Oracle

## [v7.0.4] (2016-08-18)

* Put filebeat prospector settings entirely in attributes

## [v7.0.3] (2016-08-18)

* Switch from override to default for cassandra prospector

## [v7.0.2] (2016-08-18)

* Fix the filebeat prospector syntax so that the config file actually renders

## [v7.0.1] (2016-08-18)

* Manage datastax logdir here (instead of in the wrapper)

## [v7.0.0] (2016-08-18)

* Fix the ServerSpec tests for the backups (mostly by bringing them up to date with what the code currently does)
* BREAKING: Remove all dependencies on the storage cookbook (should be handled exclusively in the wrapper)
* Do remote logging

## [v6.0.2] (2016-07-15)

* Manage cassandra init script ergo setting CASSANDRA_HEAP_DUMP_DIR and CASSANDRA_ERROR_LOG_DIR to rational values
* Remove useless heap dump setting from cassandra-env (gets overridden later)
* Update cassandra_repair serverspec test to match new cron settings

## [v6.0.1] (2016-06-17)

* Fix apt repository usage due to breaking Chef Client changes

## [v6.0.0] (2016-04-28)

* Upgrade to jdk 8

## [v5.0.9] (2016-02-25)

* Set Cassandra to restart immediately when cassandra.yaml file is sourced

## [v5.0.8] (2016-02-24)

* Fix Cassandra topology search string for Chef > 12.1.1

## [v5.0.7] (2016-02-22)

### Fixes

* Move repair jobs from 0300UTC to 1000UTC to prevent conflicts with Spark jobs

## [v5.0.6] (2016-02-16)

### Fixes

* Cassandra repair: Only error out if final retval != 0

## [v5.0.5] (2016-02-16)

### Fixes

* Cassandra repair
    - Test for success instead of failure
    - Fix null pointer in output text

### Changes

* Run repair cron job at 03:20UTC instead of 04:20UTC (so much for our calm)
* Cassandra repair: Show which retry we're on

## [v5.0.4] (2016-02-13)

### Fixes

* Check the root_dir for fs sameness before accessing work_dir

## [v5.0.3] (2016-02-13)

### Fixes

* Add more "set +/-e" flags to upload-incrementals
* Add /usr/local/bin to cron path
* Give snapshots and incrementals their own respective working directories

## [v5.0.2] (2016-02-12)

### Fixes

* Add retries to the backup and repair scripts
* Add ServerSpec tests for upload-incrementals and snapshot-cassandra

## [v5.0.1] (2016-01-06)

### Fixes

* cassandra-repair: Parse out additional stupid text from nodetool cqlsh output

## [v5.0.0] (2015-12-07)

### Changes

* Bump Cassandra to version 2.1.12

## [v4.8.14] (2015-12-14)

### Fixes

* Provide a script for uploading old snapshots

## [v4.8.13] (2015-12-14)

### Fixes

* Update serverspec test to reflect change real S3 test
* Clear out snapshots after uploading
* Check for an existing tarball before moving incremental backup files

## [v4.8.12] (2015-11-30)

### Fixes

* Repair tool: Remove goddamned carriage returns from cqlsh output

## [v4.8.11] (2015-11-30)

### Fixes

* "local" cannot be used outside of functions
* Exempt keyspaces from repair using a slightly different method

## [v4.8.10] (2015-11-25)

### Fixes

* Stop managing Cassandra backups lifecycle
* Send test backup data to real S3
* Use cassandra-stress to generate data for testing
* Fail if no OpsCenter nodes found
* Incremental Backups
    * Use find instead of ls so as to avoid grabbing regular files
    * test root_dir for same-fs state (work_dir does not exist yet)
* Repairs
    * Stop running repair in the background
    * Add preamble

## [v4.8.9] (2015-11-20)

### Fixes

* reset backup_holding_dir var in upload-incrementals

## [v4.8.8] (2015-11-20)

### Fixes

* Upload and archive old incremental backups

## [v4.8.7] (2015-11-19)

### Fixes

* Get rid of all the shell globbing in the incremental backup script

## [v4.8.6] (2015-11-18)

### Fixes

* Use find instead of rm to delete old incremental backups

## [v4.8.5] (2015-11-10)

### Fixes

* Tighten up check for existence of backup files to avoid tarring up nothing, which breaks later backups

## [v4.8.4] (2015-11-10)

### Fixes

* Check to make sure some backup files actually exist before running tar

## [v4.8.3] (2015-11-06)

### Fixes

* Do not let the backup script overwrite existing archives
* Send backup script stdout to the syslog

## [v4.8.2] (2015-11-05)

### Fixes

* Adjust `tar` arguments so that upload-incrementals does not try to delete the `backups` directory

## [v4.8.1] (2015-10-30)

### Added

* Don't spit out errors during snapshots if a keyspace doesn't have a snapshots/backups directory

## [v4.8.0] (2015-10-26)

### Added

* The two styles of backup can now be monitored separately, using separate scripts for Cronitor

### Changed

* The incremental backup now runs daily
* The snapshot backup still runs weekly, but has been moved back 12 hours to attempt to prevent it from running at the same time as an incremental backup

## [v4.7.0] (2015-10-23)

### Added

* Add incredibly prescriptive and hard-coded method to ping [Cronitor](https://cronitor.io) when the repair and backup scripts are run
    - Needs _significant_ refactoring to make it friendly to anyone else doing this other than EverTrue, but it does depend on an attribut being activated to be turned on

## [v4.6.1] (2015-10-07)

### Fixes

* Oops: s3cmd needs a verb ("put")
* Only pass the tarball file name to s3cmd, not the whole path

## [v4.6.0] (2015-10-06)

### Changes

* Enable Server Side Encryption for S3 backups
* Use s3cmd instead of boto-s3put to upload backup files (so that we can enable SSE)
* New kitchen-ec2 yaml format

### Fixes

* Move cassandra backups recipe to the end of the run list (in kitchen cloud)

attributes/default.rb
default['et_cassandra']['version']['cassandra'] = '2.1.12'
default['et_cassandra']['version']['dsc21'] =
  "#{node['et_cassandra']['version']['cassandra']}-1"
default['et_cassandra']['datastax']['version'] = '5.2.0'
default['et_cassandra']['datastax-agent']['log_dir'] = '/var/log/datastax-agent'

default['et_cassandra']['user'] = 'cassandra'
default['et_cassandra']['home'] = '/var/lib/cassandra'
default['et_cassandra']['log_dir'] = '/var/log/cassandra'
default['et_cassandra']['tarball_dir'] = '/var/lib/cassandra/tarball_dir'
default['et_cassandra']['heap_dump_dir'] = node['et_cassandra']['home']

default['et_cassandra']['limits']['nofile'] = 65_535

# Path to Cassandra configs
default['et_cassandra']['conf_path'] = '/etc/cassandra'

# As per DatStax docs re: the JNA improving memory usage
# http://www.datastax.com/documentation/cassandra/1.2/cassandra/install/installJnaDeb.html
default['et_cassandra']['packages']['jnapkg'] = 'libjna-java'

# Use jemalloc to improve malloc implementation
default['et_cassandra']['packages']['jemalloc'] = 'libjemalloc1'
default['et_cassandra']['env']['ld_library_path'] = '/usr/lib/x86_64-linux-gnu'

# We don't necessarily want all of the nodes in the cluster to restart
# upon configuration changes
default['et_cassandra']['skip_restart'] = false
default['et_cassandra']['service_action'] = [:enable, :start]

default['et_cassandra']['env']['max_heap_size'] = nil
default['et_cassandra']['env']['heap_newsize'] = nil
default['et_cassandra']['env']['heapdump_dir'] = nil
default['et_cassandra']['env']['malloc_arena_max'] = nil
default['et_cassandra']['env']['enable_gc_logging'] = false

# JVM options in this attribute as an array will be included in cassandra-env.sh
default['et_cassandra']['env']['jvm_opts'] = []

# Node discovery
default['et_cassandra']['discovery']['topo_search_str'] = 'recipes:et_cassandra\:\:default'
default['et_cassandra']['discovery']['seed_search_str'] = 'role:cassandra_seed'

# Cassandra storage config
# Values reflect the default configuration shipped w/ the Cassandra version
# specified above
# See http://wiki.apache.org/cassandra/StorageConfiguration for docs
default['et_cassandra']['config'] = {
  'cluster_name'                   => 'Test Cluster',
  'num_tokens'                     => 256,
  'hinted_handoff_enabled'         => true,
  'max_hint_window_in_ms'          => 10_800_000,
  'hinted_handoff_throttle_in_kb'  => 1024,
  'max_hints_delivery_threads'     => 2,
  'batchlog_replay_throttle_in_kb' => 1024,
  'authenticator'                  => 'AllowAllAuthenticator',
  'authorizer'                     => 'AllowAllAuthorizer',
  'permissions_validity_in_ms'     => 2000,
  'partitioner'                    => 'org.apache.cassandra.dht.Murmur3Partitioner',
  'data_file_directories'          => ["#{node['et_cassandra']['home']}/data"],
  'commitlog_directory'            => "#{node['et_cassandra']['home']}/commitlog",
  'endpoint_snitch'                => 'GossipingPropertyFileSnitch',
  'disk_failure_policy'            => 'stop',
  'commit_failure_policy'          => 'stop',
  'key_cache_size_in_mb'           => nil,
  'key_cache_save_period'          => 14_400,
  'row_cache_size_in_mb'           => 0,
  'row_cache_save_period'          => 0,
  'counter_cache_size_in_mb'       => nil,
  'counter_cache_save_period'      => 7200,
  'saved_caches_directory'         => "#{node['et_cassandra']['home']}/saved_caches",
  'commitlog_sync'                 => 'periodic',
  'commitlog_sync_period_in_ms'    => 10_000,
  'commitlog_segment_size_in_mb'   => 32,
  'concurrent_reads'                         => 32,
  'concurrent_writes'                        => 32,
  'concurrent_counter_writes'                => 32,
  'memtable_allocation_type'                 => 'heap_buffers',
  'index_summary_capacity_in_mb'             => nil,
  'index_summary_resize_interval_in_minutes' => 60,
  'trickle_fsync'                            => false,
  'trickle_fsync_interval_in_kb'             => 10_240,
  'storage_port'                             => 7000,
  'ssl_storage_port'                         => 7001,
  'listen_address'                           => node['ipaddress'],
  'start_native_transport'                   => true,
  'native_transport_port'                    => 9042,
  'start_rpc'                                => true,
  'rpc_address'                              => 'localhost',
  'rpc_port'                                 => 9160,
  'rpc_keepalive'                            => true,
  'rpc_server_type'                          => 'sync',
  'thrift_framed_transport_size_in_mb'       => 15,
  'incremental_backups'                      => true,
  'snapshot_before_compaction'               => false,
  'auto_snapshot'                            => true,
  'tombstone_warn_threshold'                 => 1000,
  'tombstone_failure_threshold'              => 100_000,
  'column_index_size_in_kb'                  => 64,
  'batch_size_warn_threshold_in_kb'          => 5,
  'compaction_throughput_mb_per_sec'         => 16,
  'sstable_preemptive_open_interval_in_mb'   => 50,
  'read_request_timeout_in_ms'               => 5000,
  'range_request_timeout_in_ms'              => 10_000,
  'write_request_timeout_in_ms'              => 2000,
  'counter_write_request_timeout_in_ms'      => 5000,
  'cas_contention_timeout_in_ms'             => 1000,
  'truncate_request_timeout_in_ms'           => 60_000,
  'request_timeout_in_ms'                    => 10_000,
  'cross_node_timeout'                       => false,
  'dynamic_snitch_update_interval_in_ms'     => 100,
  'dynamic_snitch_reset_interval_in_ms'      => 600_000,
  'dynamic_snitch_badness_threshold'         => 0.1,
  'request_scheduler'                        => 'org.apache.cassandra.scheduler.NoScheduler',
  'server_encryption_options' => {
    'internode_encryption' => 'none',
    'keystore'             => 'conf/.keystore',
    'keystore_password'    => 'cassandra',
    'truststore'           => 'conf/.truststore',
    'truststore_password'  => 'cassandra'
  },
  'client_encryption_options' => {
    'enabled'           => false,
    'keystore'          => 'conf/.keystore',
    'keystore_password' => 'cassandra'
  },
  'internode_compression' => 'all',
  'inter_dc_tcp_nodelay'  => false
}

# Java install settings
default['java']['install_flavor'] = 'openjdk'
default['java']['jdk_version'] = '8'
default['java']['java_home'] = '/usr/lib/jvm/java-7-openjdk-amd64'

# set JAVA_HOME for root so that service scripts have the correct conf
default['java']['set_etc_environment'] = true

default['et_cassandra']['cronitor']['backups_snapshot']['enabled'] = false
default['et_cassandra']['cronitor']['backups_incremental']['enabled'] = false
default['et_cassandra']['cronitor']['repairs']['enabled'] = false

default['et_cassandra']['repairs']['timeout'] = 7200
default['et_cassandra']['repairs']['retries'] = 5

# python versions
default['poise-python']['options']['get_pip_url'] = 'https://bootstrap.pypa.io/pip/2.7/get-pip.py'
default['poise-python']['options']['pip_version'] = '18.0'
default['poise-python']['options']['virtualenv_version'] = '16.7.9'
